### PR TITLE
#6974 - Add support to disable page

### LINKED ISSUE #6974: Disable pages
### Is your proposal related to a problem?

Presently, there is no way to disable pages
### Describe the solution you'd like

There should be a way to disable page so that, in the VIEWER, it is not accessible in any manner.

Homepage should not be allowed to be disabled

### PR DESCRIPTION
**Pull request by nakulnagargade**
Resolves #6974.

### CODE REVIEW

File: frontend/src/Editor/LeftSidebar/SidebarPageSelector/PageHandler.jsx [hunk 1]
```diff
@@ -137,14 +145,23 @@ export const PageHandler = ({
...
-            <SortableList.DragHandle show={isHovered} />
-          </div>
-          <div
-            className="col text-truncate font-weight-400 page-name"
-            data-cy={`pages-name-${String(page.name).toLowerCase()}`}
-          >
-            {page.name}
+            <SortableList.DragHandle show={isHovered && !isDisabled} />
           </div>
+          {!isDisabled ? (
```
Comment (johnsoncherian) [inline frontend/src/Editor/LeftSidebar/SidebarPageSelector/PageHandler.jsx:150]: Suggestion: have `disabled` text with a slightly different styling like a tag may be a better UX (How draft is in below image)
image

Comment (johnsoncherian) [reply frontend/src/Editor/LeftSidebar/SidebarPageSelector/PageHandler.jsx:150]: Or can follow the same UX of indicating with icon like how we do for hide menu.

Comment (nakulnagargade) [reply frontend/src/Editor/LeftSidebar/SidebarPageSelector/PageHandler.jsx]: Done as per new design.

File: frontend/src/Editor/LeftSidebar/SidebarPageSelector/PagehandlerMenu.jsx [hunk 2]
```diff
@@ -82,7 +90,15 @@ export const PagehandlerMenu = ({ page, darkMode, handlePageCallback, showMenu,
                   closeMenu={closeMenu}
                   callback={handlePageCallback}
                 />
-
+                <Field
```
[the remark this answers is not in the document]
Comment (nakulnagargade) [reply frontend/src/Editor/LeftSidebar/SidebarPageSelector/PagehandlerMenu.jsx:89]: Done

File: frontend/src/Editor/LeftSidebar/SidebarPageSelector/PageHandler.jsx [hunk 3]
```diff
@@ -126,36 +138,37 @@ export const PageHandler = ({
...
+            {`${page.name}`}
+            {isIconApplied && (
+              <span
+                style={{
+                  marginLeft: '8px',
+                }}
+                className="color-slate09"
+              >
+                {isHomePage && 'Home'}
+                {(isDisabled || (isDisabled && isHidden)) && 'Disabled'}
```
Comment (johnsoncherian) [inline frontend/src/Editor/LeftSidebar/SidebarPageSelector/PageHandler.jsx:169]: this condition is as good as simply checking for isDisabled right.?
`(isDisabled || (isDisabled && isHidden))` will be always equal to `isDisabled`

**Comment by johnsoncherian** [comment]
Hi @nakulnagargade overall the functionality looks good. Please check if below two points are fine.

1. We may need a different icon for disable/enable instead of settings icon. Jaseem should be able to help here.
Screenshot 2023-08-13 at 1 16 44 PM

2. Should we have a disabled page as an option for switch page.? We can check with @kavinvenkatachalam
Screenshot 2023-08-13 at 1 23 22 PM

File: frontend/src/Editor/Editor.jsx [hunk 4]
```diff
@@ -1547,6 +1571,7 @@ class EditorComponent extends React.Component {
                 renamePage={this.renamePage}
                 clonePage={this.clonePage}
                 hidePage={this.hidePage}
+                disableEnablePage={this.disableEnablePage}
```
Comment (Ckbhatia) [inline frontend/src/Editor/Editor.jsx:1574]:
```suggestion
                handlePageState={this.handlePageState}
```

File: frontend/src/Editor/LeftSidebar/SidebarPageSelector/PageHandler.jsx [hunk 5]
```diff
@@ -91,6 +97,12 @@ export const PageHandler = ({
       case 'unhide-page':
         unHidePage(page.id);
         break;
+      case 'disable-page':
+        disableEnablePage({ pageId: page.id, isDisabled: true });
+        break;
+      case 'enable-page':
+        disableEnablePage({ pageId: page.id });
```
Comment (Ckbhatia) [inline frontend/src/Editor/LeftSidebar/SidebarPageSelector/PageHandler.jsx:104]: Skipping `isDisabled: false` isn't good as it requires fellow dev to read code to understand. A bit unpredictable about handling.

File: frontend/src/Editor/LeftSidebar/SidebarPageSelector/PagehandlerMenu.jsx [hunk 6]
```diff
@@ -41,30 +49,26 @@ export const PagehandlerMenu = ({ page, darkMode, handlePageCallback, showMenu,
...
-                  id={isHidden ? 'unhide-page' : 'hide-page'}
-                  text={isHidden ? 'Unhide page' : 'Hide page'}
-                  iconSrc={`assets/images/icons/${isHidden ? 'eye' : 'eye-off'}.svg`}
-                  closeMenu={closeMenu}
-                  callback={handlePageCallback}
-                  disabled={isHome && !isHidden}
-                />
+                {!isDisabled && (
+                  <Field
+                    id={isHidden ? 'unhide-page' : 'hide-page'}
```
Comment (Ckbhatia) [inline frontend/src/Editor/LeftSidebar/SidebarPageSelector/PagehandlerMenu.jsx:64]: You can resolve, store outside JSX and use these values

Comment (Ckbhatia) [reply frontend/src/Editor/LeftSidebar/SidebarPageSelector/PagehandlerMenu.jsx:64]: Improves readability

File: frontend/src/Editor/LeftSidebar/SidebarPageSelector/index.jsx [hunk 7]
```diff
@@ -46,6 +47,11 @@ const LeftSidebarPageSelector = ({
     shallow
   );
 
+  const sortedAllPages = useMemo(
```
Comment (Ckbhatia) [inline frontend/src/Editor/LeftSidebar/SidebarPageSelector/index.jsx:50]: You can move this function outside of this component. No need to use useMemo and also `allpages` is an array.

**Review by Ckbhatia — COMMENTED** [review]
Left some comments for improvements ( optional )

File: frontend/src/Editor/LeftSidebar/SidebarPageSelector/PageHandler.jsx (hunk 3 above)
Comment (nakulnagargade) [reply frontend/src/Editor/LeftSidebar/SidebarPageSelector/PageHandler.jsx]: Yes, removed isHidden.

**Comment by nakulnagargade** [comment]
> Hi @nakulnagargade overall the functionality looks good. Please check if below two points are fine.
> 
> 1. We may need a different icon for disable/enable instead of settings icon. Jaseem should be able to help here.
> 
> Screenshot 2023-08-13 at 1 16 44 PM
> 2. Should we have a disabled page as an option for switch page.? We can check with @kavinvenkatachalam
> 
> Screenshot 2023-08-13 at 1 23 22 PM

@johnsoncherian I missed changing the icons in menubar, have corrected it. Regarding the second point, yes the disabled page shouldn't be shown in event handler, have corrected it has well.

**Review by johnsoncherian — APPROVED** [review]

**Review by johnsoncherian — DISMISSED** [review]

**Review by sherfin94 — CHANGES_REQUESTED** [review]
Hey @nakulnagargade , the error message now says 
```Attempt to switch to disabled page 'Another' blocked, redirected to home page.```

But since we are not redirecting to home page, lets change the text to 
```Attempt to switch to disabled page 'Another' blocked.```

cc: @kavinvenkatachalam , since this is just a text change, I think additional testing wont be required.

**Comment by nakulnagargade** [comment]
> Hey @nakulnagargade , the error message now says `Attempt to switch to disabled page 'Another' blocked, redirected to home page.`
> 
> But since we are not redirecting to home page, lets change the text to `Attempt to switch to disabled page 'Another' blocked.`
> 
> cc: @kavinvenkatachalam , since this is just a text change, I think additional testing wont be required.

@sherfin94 @kavinvenkatachalam Done.

**Review by sherfin94 — APPROVED** [review]

**Pull request merged**